<!DOCTYPE html>



  


<html class="theme-next mist use-motion" lang>
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform">
<meta http-equiv="Cache-Control" content="no-siteapp">
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css">







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css">

<link href="/css/main.css?v=5.1.4" rel="stylesheet" type="text/css">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.4">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.4" color="#222">





  <meta name="keywords" content="Hexo, NexT">










<meta name="description" content="0 GMATGenomic Multivariate Analysis Tools软件链接：https://github.com/chaoning/GMAT   1 ContactChao Ningningchao(at)sdau(dot)edu(dot)cnningchao91(at)gmail(dot)com   2 InstallGMAT will keep updating. Please">
<meta property="og:type" content="website">
<meta property="og:title" content="Chao Ning | 宁超">
<meta property="og:url" content="http://yoursite.com/Softwares/index.html">
<meta property="og:site_name" content="Chao Ning | 宁超">
<meta property="og:description" content="0 GMATGenomic Multivariate Analysis Tools软件链接：https://github.com/chaoning/GMAT   1 ContactChao Ningningchao(at)sdau(dot)edu(dot)cnningchao91(at)gmail(dot)com   2 InstallGMAT will keep updating. Please">
<meta property="og:locale" content="default">
<meta property="og:updated_time" content="2020-03-28T04:05:22.105Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chao Ning | 宁超">
<meta name="twitter:description" content="0 GMATGenomic Multivariate Analysis Tools软件链接：https://github.com/chaoning/GMAT   1 ContactChao Ningningchao(at)sdau(dot)edu(dot)cnningchao91(at)gmail(dot)com   2 InstallGMAT will keep updating. Please">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '5.1.4',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://yoursite.com/Softwares/">





  <title> | Chao Ning | 宁超</title>
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="default">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Chao Ning | 宁超</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-team">
          <a href="/Team" rel="section">
            
            Team
          </a>
        </li>
      
        
        <li class="menu-item menu-item-publications">
          <a href="/Publications" rel="section">
            
            Publications
          </a>
        </li>
      
        
        <li class="menu-item menu-item-softwares">
          <a href="/Softwares" rel="section">
            
            Softwares
          </a>
        </li>
      
        
        <li class="menu-item menu-item-courses">
          <a href="/Courses" rel="section">
            
            Courses
          </a>
        </li>
      
        
        <li class="menu-item menu-item-contact">
          <a href="/Contact" rel="section">
            
            Contact
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    
    
    
    <div class="post-block page">
      <header class="post-header">

	<h1 class="post-title" itemprop="name headline"></h1>



</header>

      
      
      
      <div class="post-body">
        
        
          <h1 id="0-GMAT"><a href="#0-GMAT" class="headerlink" title="0 GMAT"></a>0 GMAT</h1><p>Genomic Multivariate Analysis Tools<br>软件链接：<a href="https://github.com/chaoning/GMAT" target="_blank" rel="noopener">https://github.com/chaoning/GMAT</a>  </p>
<h1 id="1-Contact"><a href="#1-Contact" class="headerlink" title="1 Contact"></a>1 Contact</h1><p>Chao Ning<br>ningchao(at)sdau(dot)edu(dot)cn<br>ningchao91(at)gmail(dot)com  </p>
<h1 id="2-Install"><a href="#2-Install" class="headerlink" title="2 Install"></a>2 Install</h1><p>GMAT will keep updating. Please uninstall older version to obtain the latest functions. The easiest uninstall way:<br>> pip uninstall gmat</p>
<h2 id="2-1-Dependencies"><a href="#2-1-Dependencies" class="headerlink" title="2.1 Dependencies"></a>2.1 Dependencies</h2><ul>
<li>numpy&gt;=1.16.0  </li>
<li>pandas&gt;=0.19.0  </li>
<li>scipy&gt;=1.1.1  </li>
<li>cffi&gt;=1.12.0  </li>
<li>pandas_plink&gt;=2.0.0  </li>
<li>tqdm&gt;=4.43.0  </li>
</ul>
<p>We recommend using a Python distribution such as <a href="https://www.anaconda.com/distribution/" target="_blank" rel="noopener">Anaconda</a> (Python 3.7 version). This distribution can be used on Linux and Windows and is free. It is the easiest way to get all the required package dependencies. </p>
<h2 id="2-2-Quick-install"><a href="#2-2-Quick-install" class="headerlink" title="2.2 Quick install"></a>2.2 Quick install</h2><p>> pip install gmat  </p>
<h2 id="2-3-Detailed-Package-Install-Instructions"><a href="#2-3-Detailed-Package-Install-Instructions" class="headerlink" title="2.3 Detailed Package Install Instructions"></a>2.3 Detailed Package Install Instructions</h2><p>(1) Install the dependent packages<br>(2) Go to the directory of GMAT and type<br>> python setup.py install  </p>
<h1 id="3-REMMAX-function"><a href="#3-REMMAX-function" class="headerlink" title="3 REMMAX function"></a>3 REMMAX function</h1><p>Rapid Epistatic Mixed Model Association Studies</p>
<p><strong><em>Cite</em></strong>:  </p>
<ul>
<li>Dan Wang, Hui Tang, Jian-Feng Liu, Shizhong Xu, Qin Zhang and Chao Ning. Rapid Epistatic Mixed Model Association Studies by Controlling Multiple Polygenic Effects. BioRxiv, 2020. doi: <a href="https://doi.org/10.1101/2020.03.05.976498" target="_blank" rel="noopener">https://doi.org/10.1101/2020.03.05.976498</a>  </li>
<li>Chao Ning, Dan Wang, Huimin Kang, Raphael Mrode, Lei Zhou, Shizhong Xu, Jian-Feng Liu. A rapid epistatic mixed-model association analysis by linear retransformations of genomic estimated values. Bioinformatics, 2018, 34(11): 1817-1825.  </li>
</ul>
<h2 id="3-1-Format-of-the-input-file"><a href="#3-1-Format-of-the-input-file" class="headerlink" title="3.1 Format of the input file"></a>3.1 Format of the input file</h2><ul>
<li><p>Plink binary file including *.bed, *.bim and *.fam.<br>Missing genotypes are recommended to impute with Beagle or other softwares, although they will be imputed according the frequency of occurrence locus by locus.   </p>
</li>
<li><p>phenotypic file:<br>(1) Delimited by blanks or tabs;<br>(2) All individuals in the plink file must have phenotypic values. If no, please remove these individuals from the plink binary file;<br>(3) The fisrt column is the family id and the second column is the individual id. The first two columns are the same to plink fam file, but order can be different;<br>(4) The last column is the phenotypic values. <strong>Miss values are not allowed</strong>;<br>(5) The covariates (including population means) are put before the phenotypic column. A column of 1’s must be contained.<br>An example phenotypic file with four covariates (population mean, sex, age, treatmeant or untreatmeant) is as follows:<br>12659    14462    1    0    126    0    0.58<br>12659    14463    1    0    91    1    0.39<br>12659    14464    1    1    126    0    0.37<br>12659    14465    1    0    91    1    0.9<br>12659    14466    1    0    91    1    0.84<br>12659    14467    1    0    91    1    0.61<br>12659    14468    1    1    91    1    0.84  </p>
</li>
</ul>
<h2 id="3-2-Exhaustive-additive-by-addtive-epistatis"><a href="#3-2-Exhaustive-additive-by-addtive-epistatis" class="headerlink" title="3.2 Exhaustive additive by addtive epistatis"></a>3.2 Exhaustive additive by addtive epistatis</h2><p>Data: Mouse data in directory of GMAT/examples/data/mouse    </p>
<h3 id="3-2-1-Include-additive-and-additive-by-additive-genomic-relationship-matrix"><a href="#3-2-1-Include-additive-and-additive-by-additive-genomic-relationship-matrix" class="headerlink" title="3.2.1 Include additive and additive by additive genomic relationship matrix"></a>3.2.1 Include additive and additive by additive genomic relationship matrix</h3><h4 id="1-Exact-test-for-small-data"><a href="#1-Exact-test-for-small-data" class="headerlink" title="(1)  Exact test (for small data)"></a>(1)  Exact test (for small data)</h4><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> logging</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> gmat.gmatrix <span class="keyword">import</span> agmat, dgmat_as</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.design_matrix <span class="keyword">import</span> design_matrix_wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.uvlmm_varcom <span class="keyword">import</span> wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_cpu, remma_epiAA_cpu_parallel</span><br><span class="line"><span class="keyword">from</span> gmat.remma.annotation <span class="keyword">import</span> annotation_snp_pos</span><br><span class="line">logging.basicConfig(level=logging.INFO)</span><br><span class="line"></span><br><span class="line">pheno_file = <span class="string">'pheno'</span>  <span class="comment"># phenotypic file</span></span><br><span class="line">bed_file = <span class="string">'plink'</span>  <span class="comment"># the prefix for the plink binary file</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 1: Calculate the genomic relationship matrix</span></span><br><span class="line">agmat_lst = agmat(bed_file, inv=<span class="literal">False</span>) <span class="comment"># additive genomic relationship matrix</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 2: Prepare the phenotypic vector (y), designed matrix for fixed effects (xmat) and designed matrix for random effects (zmat)</span></span><br><span class="line">y, xmat, zmat = design_matrix_wemai_multi_gmat(pheno_file, bed_file)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 3: Estimate the variances</span></span><br><span class="line">gmat_lst = [agmat_lst[<span class="number">0</span>], agmat_lst[<span class="number">0</span>]*agmat_lst[<span class="number">0</span>]]  <span class="comment"># agmat_lst[0]*agmat_lst[0] is the additive by additive genomic relationship matrix</span></span><br><span class="line">var_com_a_axa = wemai_multi_gmat(y, xmat, zmat, gmat_lst)</span><br><span class="line">print(var_com_a_axa)  <span class="comment"># a list： [0] addtive variance; [1] additive by additive variance; [2] residual variance</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 4: Test</span></span><br><span class="line">remma_epiAA_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_a_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 5: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_a_axa'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h4 id="2-Parallel-exact-test-for-small-data"><a href="#2-Parallel-exact-test-for-small-data" class="headerlink" title="(2) Parallel exact test  (for small data)"></a>(2) Parallel exact test  (for small data)</h4><p>Analysis can be subdivided with remma_epiAA_cpu_parallel and run parallelly on different machines.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Step 1-3 is same to the above</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 4: parallel test. Write the codes in separate scripts and run separately.</span></span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_cpu_parallel</span><br><span class="line">remma_epiAA_cpu_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">1</span>], </span><br><span class="line">                         p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_a_axa'</span>)</span><br><span class="line">remma_epiAA_cpu_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">2</span>], </span><br><span class="line">                         p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_a_axa'</span>)</span><br><span class="line">remma_epiAA_cpu_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">3</span>], </span><br><span class="line">                         p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_a_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 5: Merge files 'epiAA_a_axa.1', 'epiAA_a_axa.2' and 'epiAA_a_axa.3' with the following codes.</span></span><br><span class="line">prefix = <span class="string">'epiAA_a_axa'</span></span><br><span class="line">parallel_num = <span class="number">3</span>  <span class="comment"># the number of parallels</span></span><br><span class="line"><span class="keyword">with</span> open(prefix + <span class="string">".merge"</span>, <span class="string">'w'</span>) <span class="keyword">as</span> fout:</span><br><span class="line">    <span class="keyword">with</span> open(prefix + <span class="string">'.1'</span>) <span class="keyword">as</span> fin:</span><br><span class="line">        head_line = fin.readline()</span><br><span class="line">        fout.write(head_line)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>, <span class="number">4</span>):</span><br><span class="line">        <span class="keyword">with</span> open(prefix + <span class="string">'.'</span> + str(i)) <span class="keyword">as</span> fin:</span><br><span class="line">            head_line = fin.readline()</span><br><span class="line">            <span class="keyword">for</span> line <span class="keyword">in</span> fin:</span><br><span class="line">                fout.write(line)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 6: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_a_axa.merge'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h4 id="3-approximate-test-recommended-for-big-data"><a href="#3-approximate-test-recommended-for-big-data" class="headerlink" title="(3) approximate test (recommended for big data)"></a>(3) approximate test (recommended for big data)</h4><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> logging</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> scipy.stats <span class="keyword">import</span> chi2</span><br><span class="line"><span class="keyword">from</span> gmat.gmatrix <span class="keyword">import</span> agmat, dgmat_as</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.design_matrix <span class="keyword">import</span> design_matrix_wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.uvlmm_varcom <span class="keyword">import</span> wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.remma.random_pair <span class="keyword">import</span> random_pair</span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_pair_cpu, remma_epiAA_eff_cpu_c, remma_epiAA_eff_cpu_c_parallel</span><br><span class="line"><span class="keyword">from</span> gmat.remma.annotation <span class="keyword">import</span> annotation_snp_pos</span><br><span class="line">logging.basicConfig(level=logging.INFO)</span><br><span class="line"></span><br><span class="line">pheno_file = <span class="string">'pheno'</span>  <span class="comment"># phenotypic file</span></span><br><span class="line">bed_file = <span class="string">'plink'</span>  <span class="comment"># the prefix for the plink binary file</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 1-3 is same to exact test</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 4: Randomly select 100,000 SNP pairs</span></span><br><span class="line">snp_df = pd.read_csv(bed_file + <span class="string">'.bim'</span>, header=<span class="literal">None</span>, sep=<span class="string">'\s+'</span>)</span><br><span class="line">num_snp = snp_df.shape[<span class="number">0</span>]  <span class="comment"># the number of snp</span></span><br><span class="line">random_pair(num_snp, out_file=<span class="string">'random_pair'</span>, num_pair=<span class="number">100000</span>, num_each_pair=<span class="number">5000</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># step 5: Test these 100,000 SNP pairs</span></span><br><span class="line"><span class="comment"># note: set p_cut=1 to save all the results</span></span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_pair_file=<span class="string">"random_pair"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_random_a_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># step 6: Calculate the median of variances for estimated epistatic SNP effects</span></span><br><span class="line">res_df = pd.read_csv(<span class="string">'epiAA_pair_random_a_axa'</span>, header=<span class="number">0</span>, sep=<span class="string">'\s+'</span>)</span><br><span class="line">print(np.median(res_df[<span class="string">'p'</span>]))  <span class="comment"># P value close to 0.5. It means type I error controlled well</span></span><br><span class="line">var_median = np.median(res_df[<span class="string">'var'</span>])  <span class="comment"># median of variances for estimated epistatic SNP effects</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># step 7: Screen the effects and select top SNP pairs based on approximate test. </span></span><br><span class="line"><span class="comment"># Use the above median of variances as the approximate values (var_app = var_median)</span></span><br><span class="line">remma_epiAA_eff_cpu_c(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, var_app=var_median, </span><br><span class="line">                      p_cut=<span class="number">1e-05</span>, out_file=<span class="string">'epiAA_eff_c_a_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 8: Calculate exact p values for top SNP pairs</span></span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_a_axa"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_a_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 9: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_pair_res_a_axa'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h4 id="4-Parallel-approximate-test-recommended-for-big-data"><a href="#4-Parallel-approximate-test-recommended-for-big-data" class="headerlink" title="(4) Parallel approximate test (recommended for big data)"></a>(4) Parallel approximate test (recommended for big data)</h4><p>Analysis can be subdivided with remma_epiAA_eff_cpu_c_parallel and run parallelly on different machines.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Step 1-6 is same to the above</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 7: parallel test. Write the codes in separate scripts and run separately.</span></span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_eff_cpu_c_parallel</span><br><span class="line">remma_epiAA_eff_cpu_c_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">1</span>], </span><br><span class="line">                               var_app=var_median, p_cut=<span class="number">1.0e-5</span>, out_file=<span class="string">'epiAA_eff_c_a_axa'</span>)</span><br><span class="line">remma_epiAA_eff_cpu_c_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">2</span>], </span><br><span class="line">                               var_app=var_median, p_cut=<span class="number">1.0e-5</span>, out_file=<span class="string">'epiAA_eff_c_a_axa'</span>)</span><br><span class="line">remma_epiAA_eff_cpu_c_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">3</span>], </span><br><span class="line">                               var_app=var_median, p_cut=<span class="number">1.0e-5</span>, out_file=<span class="string">'epiAA_eff_c_a_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 8: Calculate exact p values for top SNP pairs</span></span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_a_axa.1"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_a_axa.1'</span>)</span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_a_axa.2"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_a_axa.2'</span>)</span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_axa, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_a_axa.3"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_a_axa.3'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 9: Merge files 'epiAA_pair_res_a_axa.1', 'epiAA_pair_res_a_axa.2' and 'epiAA_pair_res_a_axa.3' </span></span><br><span class="line"><span class="comment"># with the following codes.</span></span><br><span class="line">prefix = <span class="string">'epiAA_pair_res_a_axa'</span></span><br><span class="line">parallel_num = <span class="number">3</span>  <span class="comment"># the number of parallels</span></span><br><span class="line"><span class="keyword">with</span> open(prefix + <span class="string">".merge"</span>, <span class="string">'w'</span>) <span class="keyword">as</span> fout:</span><br><span class="line">    <span class="keyword">with</span> open(prefix + <span class="string">'.1'</span>) <span class="keyword">as</span> fin:</span><br><span class="line">        head_line = fin.readline()</span><br><span class="line">        fout.write(head_line)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>, <span class="number">4</span>):</span><br><span class="line">        <span class="keyword">with</span> open(prefix + <span class="string">'.'</span> + str(i)) <span class="keyword">as</span> fin:</span><br><span class="line">            head_line = fin.readline()</span><br><span class="line">            <span class="keyword">for</span> line <span class="keyword">in</span> fin:</span><br><span class="line">                fout.write(line)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 10: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_pair_res_a_axa.merge'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h3 id="3-2-2-Include-additive-dominance-and-additive-by-additive-genomic-relationship-matrix"><a href="#3-2-2-Include-additive-dominance-and-additive-by-additive-genomic-relationship-matrix" class="headerlink" title="3.2.2 Include additive, dominance and additive by additive genomic relationship matrix"></a>3.2.2 Include additive, dominance and additive by additive genomic relationship matrix</h3><h4 id="1-Exact-test-for-small-data-1"><a href="#1-Exact-test-for-small-data-1" class="headerlink" title="(1) Exact test (for small data)"></a>(1) Exact test (for small data)</h4><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> logging</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> gmat.gmatrix <span class="keyword">import</span> agmat, dgmat_as</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.design_matrix <span class="keyword">import</span> design_matrix_wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.uvlmm_varcom <span class="keyword">import</span> wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_cpu, remma_epiAA_cpu_parallel</span><br><span class="line"><span class="keyword">from</span> gmat.remma.annotation <span class="keyword">import</span> annotation_snp_pos</span><br><span class="line">logging.basicConfig(level=logging.INFO)</span><br><span class="line"></span><br><span class="line">pheno_file = <span class="string">'pheno'</span></span><br><span class="line">bed_file = <span class="string">'plink'</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 1: Calculate the genomic relationship matrix</span></span><br><span class="line">agmat_lst = agmat(bed_file, inv=<span class="literal">False</span>) <span class="comment"># additive genomic relationship matrix</span></span><br><span class="line">dgmat_lst = dgmat_as(bed_file, inv=<span class="literal">False</span>) <span class="comment"># dominace genomic relationship matrix</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 2: Prepare the phenotypic vector (y), designed matrix for fixed effects (xmat) and designed matrix for random effects (zmat)</span></span><br><span class="line">y, xmat, zmat = design_matrix_wemai_multi_gmat(pheno_file, bed_file)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 3: Estimate the variances</span></span><br><span class="line">gmat_lst = [agmat_lst[<span class="number">0</span>], dgmat_lst[<span class="number">0</span>], agmat_lst[<span class="number">0</span>]*agmat_lst[<span class="number">0</span>]]  <span class="comment"># agmat_lst[0]*agmat_lst[0] is the additive by additive genomic relationship matrix</span></span><br><span class="line">var_com_a_d_axa = wemai_multi_gmat(y, xmat, zmat, gmat_lst)</span><br><span class="line">print(var_com_a_d_axa)  <span class="comment"># a list： [0] addtive variance; [1] dominace variance [2] additive by additive variance; [3] residual variance</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 4: Test</span></span><br><span class="line">remma_epiAA_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_a_d_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 5: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_a_d_axa'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h4 id="2-Parallel-exact-test-for-small-data-1"><a href="#2-Parallel-exact-test-for-small-data-1" class="headerlink" title="(2) Parallel exact test (for small data)"></a>(2) Parallel exact test (for small data)</h4><p>Analysis can be subdivided with remma_epiAA_cpu_parallel and run parallelly on different machines.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Step 1-3 is same to the above</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 4: parallel test. Write the codes in separate scripts and run separately.</span></span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_cpu_parallel</span><br><span class="line">remma_epiAA_cpu_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">1</span>], </span><br><span class="line">                         p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_a_d_axa'</span>)</span><br><span class="line">remma_epiAA_cpu_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">2</span>], </span><br><span class="line">                         p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_a_d_axa'</span>)</span><br><span class="line">remma_epiAA_cpu_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">3</span>], </span><br><span class="line">                         p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_a_d_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 5: Merge files 'epiAA_a_d_axa.1', 'epiAA_a_d_axa.2' and 'epiAA_a_d_axa.3' with the following codes.</span></span><br><span class="line">prefix = <span class="string">'epiAA_a_d_axa'</span></span><br><span class="line">parallel_num = <span class="number">3</span>  <span class="comment"># the number of parallels</span></span><br><span class="line"><span class="keyword">with</span> open(prefix + <span class="string">".merge"</span>, <span class="string">'w'</span>) <span class="keyword">as</span> fout:</span><br><span class="line">    <span class="keyword">with</span> open(prefix + <span class="string">'.1'</span>) <span class="keyword">as</span> fin:</span><br><span class="line">        head_line = fin.readline()</span><br><span class="line">        fout.write(head_line)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>, <span class="number">4</span>):</span><br><span class="line">        <span class="keyword">with</span> open(prefix + <span class="string">'.'</span> + str(i)) <span class="keyword">as</span> fin:</span><br><span class="line">            head_line = fin.readline()</span><br><span class="line">            <span class="keyword">for</span> line <span class="keyword">in</span> fin:</span><br><span class="line">                fout.write(line)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 6: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_a_d_axa.merge'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h4 id="3-approximate-test-recommended-for-big-data-1"><a href="#3-approximate-test-recommended-for-big-data-1" class="headerlink" title="(3) approximate test (recommended for big data)"></a>(3) approximate test (recommended for big data)</h4><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> logging</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> scipy.stats <span class="keyword">import</span> chi2</span><br><span class="line"><span class="keyword">from</span> gmat.gmatrix <span class="keyword">import</span> agmat, dgmat_as</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.design_matrix <span class="keyword">import</span> design_matrix_wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.uvlmm_varcom <span class="keyword">import</span> wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.remma.random_pair <span class="keyword">import</span> random_pair</span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_pair_cpu, remma_epiAA_eff_cpu_c, remma_epiAA_eff_cpu_c_parallel</span><br><span class="line"><span class="keyword">from</span> gmat.remma.annotation <span class="keyword">import</span> annotation_snp_pos</span><br><span class="line">logging.basicConfig(level=logging.INFO)</span><br><span class="line"></span><br><span class="line">pheno_file = <span class="string">'pheno'</span>  <span class="comment"># phenotypic file</span></span><br><span class="line">bed_file = <span class="string">'plink'</span>  <span class="comment"># the prefix for the plink binary file</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 1-3 is same to exact test</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 4: Randomly select 100,000 SNP pairs</span></span><br><span class="line">snp_df = pd.read_csv(bed_file + <span class="string">'.bim'</span>, header=<span class="literal">None</span>, sep=<span class="string">'\s+'</span>)</span><br><span class="line">num_snp = snp_df.shape[<span class="number">0</span>]  <span class="comment"># the number of snp</span></span><br><span class="line">random_pair(num_snp, out_file=<span class="string">'random_pair'</span>, num_pair=<span class="number">100000</span>, num_each_pair=<span class="number">5000</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># step 5: Test these 100,000 SNP pairs</span></span><br><span class="line"><span class="comment"># note: set p_cut=1 to save all the results</span></span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, snp_pair_file=<span class="string">"random_pair"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_random_a_d_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># step 6: Calculate the median of variances for estimated epistatic SNP effects</span></span><br><span class="line">res_df = pd.read_csv(<span class="string">'epiAA_pair_random_a_d_axa'</span>, header=<span class="number">0</span>, sep=<span class="string">'\s+'</span>)</span><br><span class="line">print(np.median(res_df[<span class="string">'p'</span>]))  <span class="comment"># P value close to 0.5. It means type I error controlled well</span></span><br><span class="line">var_median = np.median(res_df[<span class="string">'var'</span>])  <span class="comment"># median of variances for estimated epistatic SNP effects</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># step 7: Screen the effects and select top SNP pairs based on approximate test. </span></span><br><span class="line"><span class="comment"># Use the above median of variances as the approximate values (var_app = var_median)</span></span><br><span class="line">remma_epiAA_eff_cpu_c(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, var_app=var_median, </span><br><span class="line">                      p_cut=<span class="number">1e-05</span>, out_file=<span class="string">'epiAA_eff_c_a_d_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 8: Calculate exact p values for top SNP pairs</span></span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_a_d_axa"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_a_d_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 9: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_pair_res_a_d_axa'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h4 id="4-Parallel-approximate-test-recommended-for-big-data-1"><a href="#4-Parallel-approximate-test-recommended-for-big-data-1" class="headerlink" title="(4) Parallel approximate test (recommended for big data)"></a>(4) Parallel approximate test (recommended for big data)</h4><p>Analysis can be subdivided with remma_epiAA_eff_cpu_c_parallel and run parallelly on different machines.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Step 1-6 is same to the above</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 7: parallel test. Write the codes in separate scripts and run separately.</span></span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_eff_cpu_c_parallel</span><br><span class="line">remma_epiAA_eff_cpu_c_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">1</span>], </span><br><span class="line">                               var_app=var_median, p_cut=<span class="number">1.0e-5</span>, out_file=<span class="string">'epiAA_eff_c_a_d_axa'</span>)</span><br><span class="line">remma_epiAA_eff_cpu_c_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">2</span>], </span><br><span class="line">                               var_app=var_median, p_cut=<span class="number">1.0e-5</span>, out_file=<span class="string">'epiAA_eff_c_a_d_axa'</span>)</span><br><span class="line">remma_epiAA_eff_cpu_c_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">3</span>], </span><br><span class="line">                               var_app=var_median, p_cut=<span class="number">1.0e-5</span>, out_file=<span class="string">'epiAA_eff_c_a_d_axa'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 8: Calculate exact p values for top SNP pairs</span></span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_a_d_axa.1"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_a_d_axa.1'</span>)</span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_a_d_axa.2"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_a_d_axa.2'</span>)</span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_a_d_axa, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_a_d_axa.3"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_a_d_axa.3'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 9: Merge files 'epiAA_pair_res_a_d_axa.1', 'epiAA_pair_res_a_d_axa.2' and 'epiAA_pair_res_a_d_axa.3' </span></span><br><span class="line"><span class="comment"># with the following codes.</span></span><br><span class="line">prefix = <span class="string">'epiAA_pair_res_a_d_axa'</span></span><br><span class="line">parallel_num = <span class="number">3</span>  <span class="comment"># the number of parallels</span></span><br><span class="line"><span class="keyword">with</span> open(prefix + <span class="string">".merge"</span>, <span class="string">'w'</span>) <span class="keyword">as</span> fout:</span><br><span class="line">    <span class="keyword">with</span> open(prefix + <span class="string">'.1'</span>) <span class="keyword">as</span> fin:</span><br><span class="line">        head_line = fin.readline()</span><br><span class="line">        fout.write(head_line)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>, <span class="number">4</span>):</span><br><span class="line">        <span class="keyword">with</span> open(prefix + <span class="string">'.'</span> + str(i)) <span class="keyword">as</span> fin:</span><br><span class="line">            head_line = fin.readline()</span><br><span class="line">            <span class="keyword">for</span> line <span class="keyword">in</span> fin:</span><br><span class="line">                fout.write(line)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 10: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_pair_res_a_d_axa.merge'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h3 id="3-2-3-Include-additive-dominance-and-three-kinds-of-epistatic-genomic-relationship-matrix"><a href="#3-2-3-Include-additive-dominance-and-three-kinds-of-epistatic-genomic-relationship-matrix" class="headerlink" title="3.2.3 Include additive, dominance and three kinds of epistatic genomic relationship matrix"></a>3.2.3 Include additive, dominance and three kinds of epistatic genomic relationship matrix</h3><p>additive, dominance, additive by additive, additive by dominance and dominance by dominance genomic relationship matrix   </p>
<h4 id="1-Exact-test-for-small-data-2"><a href="#1-Exact-test-for-small-data-2" class="headerlink" title="(1) Exact test (for small data)"></a>(1) Exact test (for small data)</h4><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> logging</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> gmat.gmatrix <span class="keyword">import</span> agmat, dgmat_as</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.design_matrix <span class="keyword">import</span> design_matrix_wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.uvlmm_varcom <span class="keyword">import</span> wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_cpu, remma_epiAA_cpu_parallel</span><br><span class="line"><span class="keyword">from</span> gmat.remma.annotation <span class="keyword">import</span> annotation_snp_pos</span><br><span class="line">logging.basicConfig(level=logging.INFO)</span><br><span class="line"></span><br><span class="line">pheno_file = <span class="string">'pheno'</span></span><br><span class="line">bed_file = <span class="string">'plink'</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 1: Calculate the genomic relationship matrix</span></span><br><span class="line">agmat_lst = agmat(bed_file, inv=<span class="literal">False</span>) <span class="comment"># additive genomic relationship matrix</span></span><br><span class="line">dgmat_lst = dgmat_as(bed_file, inv=<span class="literal">False</span>) <span class="comment"># dominace genomic relationship matrix</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 2: Prepare the phenotypic vector (y), designed matrix for fixed effects (xmat) and designed matrix for random effects (zmat)</span></span><br><span class="line">y, xmat, zmat = design_matrix_wemai_multi_gmat(pheno_file, bed_file)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 3: Estimate the variances</span></span><br><span class="line"><span class="comment"># agmat_lst[0]*agmat_lst[0] is the additive by additive genomic relationship matrix</span></span><br><span class="line"><span class="comment"># agmat_lst[0]*dgmat_lst[0] is the additive by dominance genomic relationship matrix</span></span><br><span class="line"><span class="comment"># dgmat_lst[0]*dgmat_lst[0] is the dominance by dominance genomic relationship matrix</span></span><br><span class="line">gmat_lst = [agmat_lst[<span class="number">0</span>], dgmat_lst[<span class="number">0</span>], agmat_lst[<span class="number">0</span>]*agmat_lst[<span class="number">0</span>], agmat_lst[<span class="number">0</span>]*dgmat_lst[<span class="number">0</span>], dgmat_lst[<span class="number">0</span>]*dgmat_lst[<span class="number">0</span>]]  </span><br><span class="line"></span><br><span class="line">var_com_all = wemai_multi_gmat(y, xmat, zmat, gmat_lst)</span><br><span class="line">print(var_com_all)  <span class="comment"># a list： [0] addtive variance; [1] dominace variance [2] additive by additive variance; [3] residual variance</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 4: Test</span></span><br><span class="line">remma_epiAA_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_all'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 5: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_all'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h4 id="2-Parallel-exact-test-for-small-data-2"><a href="#2-Parallel-exact-test-for-small-data-2" class="headerlink" title="(2) Parallel exact test (for small data)"></a>(2) Parallel exact test (for small data)</h4><p>Analysis can be subdivided with remma_epiAA_cpu_parallel and run parallelly on different machines.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Step 1-3 is same to the above</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 4: parallel test. Write the codes in separate scripts and run separately.</span></span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_cpu_parallel</span><br><span class="line">remma_epiAA_cpu_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">1</span>], </span><br><span class="line">                         p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_all'</span>)</span><br><span class="line">remma_epiAA_cpu_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">2</span>], </span><br><span class="line">                         p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_all'</span>)</span><br><span class="line">remma_epiAA_cpu_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">3</span>], </span><br><span class="line">                         p_cut=<span class="number">0.0001</span>, out_file=<span class="string">'epiAA_all'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 5: Merge files 'epiAA_a_d_axa.1', 'epiAA_a_d_axa.2' and 'epiAA_a_d_axa.3' with the following codes.</span></span><br><span class="line">prefix = <span class="string">'epiAA_all'</span></span><br><span class="line">parallel_num = <span class="number">3</span>  <span class="comment"># the number of parallels</span></span><br><span class="line"><span class="keyword">with</span> open(prefix + <span class="string">".merge"</span>, <span class="string">'w'</span>) <span class="keyword">as</span> fout:</span><br><span class="line">    <span class="keyword">with</span> open(prefix + <span class="string">'.1'</span>) <span class="keyword">as</span> fin:</span><br><span class="line">        head_line = fin.readline()</span><br><span class="line">        fout.write(head_line)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>, parallel_num+<span class="number">1</span>):</span><br><span class="line">        <span class="keyword">with</span> open(prefix + <span class="string">'.'</span> + str(i)) <span class="keyword">as</span> fin:</span><br><span class="line">            head_line = fin.readline()</span><br><span class="line">            <span class="keyword">for</span> line <span class="keyword">in</span> fin:</span><br><span class="line">                fout.write(line)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 6: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_all.merge'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h4 id="3-approximate-test-recommended-for-big-data-2"><a href="#3-approximate-test-recommended-for-big-data-2" class="headerlink" title="(3) approximate test (recommended for big data)"></a>(3) approximate test (recommended for big data)</h4><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> logging</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> scipy.stats <span class="keyword">import</span> chi2</span><br><span class="line"><span class="keyword">from</span> gmat.gmatrix <span class="keyword">import</span> agmat, dgmat_as</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.design_matrix <span class="keyword">import</span> design_matrix_wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.uvlmm.uvlmm_varcom <span class="keyword">import</span> wemai_multi_gmat</span><br><span class="line"><span class="keyword">from</span> gmat.remma.random_pair <span class="keyword">import</span> random_pair</span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_pair_cpu, remma_epiAA_eff_cpu_c, remma_epiAA_eff_cpu_c_parallel</span><br><span class="line"><span class="keyword">from</span> gmat.remma.annotation <span class="keyword">import</span> annotation_snp_pos</span><br><span class="line">logging.basicConfig(level=logging.INFO)</span><br><span class="line"></span><br><span class="line">pheno_file = <span class="string">'pheno'</span>  <span class="comment"># phenotypic file</span></span><br><span class="line">bed_file = <span class="string">'plink'</span>  <span class="comment"># the prefix for the plink binary file</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 1-3 is same to exact test</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 4: Randomly select 100,000 SNP pairs</span></span><br><span class="line">snp_df = pd.read_csv(bed_file + <span class="string">'.bim'</span>, header=<span class="literal">None</span>, sep=<span class="string">'\s+'</span>)</span><br><span class="line">num_snp = snp_df.shape[<span class="number">0</span>]  <span class="comment"># the number of snp</span></span><br><span class="line">random_pair(num_snp, out_file=<span class="string">'random_pair'</span>, num_pair=<span class="number">100000</span>, num_each_pair=<span class="number">5000</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># step 5: Test these 100,000 SNP pairs</span></span><br><span class="line"><span class="comment"># note: set p_cut=1 to save all the results</span></span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, snp_pair_file=<span class="string">"random_pair"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_random_all'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># step 6: Calculate the median of variances for estimated epistatic SNP effects</span></span><br><span class="line">res_df = pd.read_csv(<span class="string">'epiAA_pair_random_all'</span>, header=<span class="number">0</span>, sep=<span class="string">'\s+'</span>)</span><br><span class="line">print(np.median(res_df[<span class="string">'p'</span>]))  <span class="comment"># P value close to 0.5. It means type I error controlled well</span></span><br><span class="line">var_median = np.median(res_df[<span class="string">'var'</span>])  <span class="comment"># median of variances for estimated epistatic SNP effects</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># step 7: Screen the effects and select top SNP pairs based on approximate test. </span></span><br><span class="line"><span class="comment"># Use the above median of variances as the approximate values (var_app = var_median)</span></span><br><span class="line">remma_epiAA_eff_cpu_c(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, var_app=var_median, </span><br><span class="line">                      p_cut=<span class="number">1e-05</span>, out_file=<span class="string">'epiAA_eff_c_all'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 8: Calculate exact p values for top SNP pairs</span></span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_all"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_all'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 9: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_pair_res_all'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>
<h4 id="4-Parallel-approximate-test-recommended-for-big-data-2"><a href="#4-Parallel-approximate-test-recommended-for-big-data-2" class="headerlink" title="(4) Parallel approximate test (recommended for big data)"></a>(4) Parallel approximate test (recommended for big data)</h4><p>Analysis can be subdivided with remma_epiAA_eff_cpu_c_parallel and run parallelly on different machines.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Step 1-6 is same to the above</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 7: parallel test. Write the codes in separate scripts and run separately.</span></span><br><span class="line"><span class="keyword">from</span> gmat.remma.remma_cpu.remma_epiAA_cpu <span class="keyword">import</span> remma_epiAA_eff_cpu_c_parallel</span><br><span class="line">remma_epiAA_eff_cpu_c_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">1</span>], </span><br><span class="line">                               var_app=var_median, p_cut=<span class="number">1.0e-5</span>, out_file=<span class="string">'epiAA_eff_c_all'</span>)</span><br><span class="line">remma_epiAA_eff_cpu_c_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">2</span>], </span><br><span class="line">                               var_app=var_median, p_cut=<span class="number">1.0e-5</span>, out_file=<span class="string">'epiAA_eff_c_all'</span>)</span><br><span class="line">remma_epiAA_eff_cpu_c_parallel(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, parallel=[<span class="number">3</span>,<span class="number">3</span>], </span><br><span class="line">                               var_app=var_median, p_cut=<span class="number">1.0e-5</span>, out_file=<span class="string">'epiAA_eff_c_all'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 8: Calculate exact p values for top SNP pairs</span></span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_all.1"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_all.1'</span>)</span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_all.2"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_all.2'</span>)</span><br><span class="line">remma_epiAA_pair_cpu(y, xmat, zmat, gmat_lst, var_com=var_com_all, bed_file=bed_file, snp_pair_file=<span class="string">"epiAA_eff_c_all.3"</span>, </span><br><span class="line">                     max_test_pair=<span class="number">50000</span>, p_cut=<span class="number">1</span>, out_file=<span class="string">'epiAA_pair_res_all.3'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 9: Merge files 'epiAA_pair_res_all.1', 'epiAA_pair_res_all.2' and 'epiAA_pair_res_all.3' </span></span><br><span class="line"><span class="comment"># with the following codes.</span></span><br><span class="line">prefix = <span class="string">'epiAA_pair_res_all'</span></span><br><span class="line">parallel_num = <span class="number">3</span>  <span class="comment"># the number of parallels</span></span><br><span class="line"><span class="keyword">with</span> open(prefix + <span class="string">".merge"</span>, <span class="string">'w'</span>) <span class="keyword">as</span> fout:</span><br><span class="line">    <span class="keyword">with</span> open(prefix + <span class="string">'.1'</span>) <span class="keyword">as</span> fin:</span><br><span class="line">        head_line = fin.readline()</span><br><span class="line">        fout.write(head_line)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>, parallel_num+<span class="number">1</span>):</span><br><span class="line">        <span class="keyword">with</span> open(prefix + <span class="string">'.'</span> + str(i)) <span class="keyword">as</span> fin:</span><br><span class="line">            head_line = fin.readline()</span><br><span class="line">            <span class="keyword">for</span> line <span class="keyword">in</span> fin:</span><br><span class="line">                fout.write(line)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Step 10: Select top SNPs and add the SNP position</span></span><br><span class="line">res_file = <span class="string">'epiAA_pair_res_all.merge'</span>  <span class="comment"># result file</span></span><br><span class="line">annotation_snp_pos(res_file, bed_file, p_cut=<span class="number">1.0e-5</span>)</span><br></pre></td></tr></table></figure>

        
      </div>
      
      
      
    </div>
    
    
    
  </div>


          </div>
          


          

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">Chao Ning</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives">
              
                  <span class="site-state-item-count">1</span>
                  <span class="site-state-item-name">posts</span>
                </a>
              </div>
            

            

            

          </nav>

          

          

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#0-GMAT"><span class="nav-number">1.</span> <span class="nav-text">0 GMAT</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#1-Contact"><span class="nav-number">2.</span> <span class="nav-text">1 Contact</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#2-Install"><span class="nav-number">3.</span> <span class="nav-text">2 Install</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#2-1-Dependencies"><span class="nav-number">3.1.</span> <span class="nav-text">2.1 Dependencies</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-2-Quick-install"><span class="nav-number">3.2.</span> <span class="nav-text">2.2 Quick install</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-3-Detailed-Package-Install-Instructions"><span class="nav-number">3.3.</span> <span class="nav-text">2.3 Detailed Package Install Instructions</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#3-REMMAX-function"><span class="nav-number">4.</span> <span class="nav-text">3 REMMAX function</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#3-1-Format-of-the-input-file"><span class="nav-number">4.1.</span> <span class="nav-text">3.1 Format of the input file</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#3-2-Exhaustive-additive-by-addtive-epistatis"><span class="nav-number">4.2.</span> <span class="nav-text">3.2 Exhaustive additive by addtive epistatis</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#3-2-1-Include-additive-and-additive-by-additive-genomic-relationship-matrix"><span class="nav-number">4.2.1.</span> <span class="nav-text">3.2.1 Include additive and additive by additive genomic relationship matrix</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#1-Exact-test-for-small-data"><span class="nav-number">4.2.1.1.</span> <span class="nav-text">(1)  Exact test (for small data)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-Parallel-exact-test-for-small-data"><span class="nav-number">4.2.1.2.</span> <span class="nav-text">(2) Parallel exact test  (for small data)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-approximate-test-recommended-for-big-data"><span class="nav-number">4.2.1.3.</span> <span class="nav-text">(3) approximate test (recommended for big data)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-Parallel-approximate-test-recommended-for-big-data"><span class="nav-number">4.2.1.4.</span> <span class="nav-text">(4) Parallel approximate test (recommended for big data)</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-2-2-Include-additive-dominance-and-additive-by-additive-genomic-relationship-matrix"><span class="nav-number">4.2.2.</span> <span class="nav-text">3.2.2 Include additive, dominance and additive by additive genomic relationship matrix</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#1-Exact-test-for-small-data-1"><span class="nav-number">4.2.2.1.</span> <span class="nav-text">(1) Exact test (for small data)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-Parallel-exact-test-for-small-data-1"><span class="nav-number">4.2.2.2.</span> <span class="nav-text">(2) Parallel exact test (for small data)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-approximate-test-recommended-for-big-data-1"><span class="nav-number">4.2.2.3.</span> <span class="nav-text">(3) approximate test (recommended for big data)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-Parallel-approximate-test-recommended-for-big-data-1"><span class="nav-number">4.2.2.4.</span> <span class="nav-text">(4) Parallel approximate test (recommended for big data)</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-2-3-Include-additive-dominance-and-three-kinds-of-epistatic-genomic-relationship-matrix"><span class="nav-number">4.2.3.</span> <span class="nav-text">3.2.3 Include additive, dominance and three kinds of epistatic genomic relationship matrix</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#1-Exact-test-for-small-data-2"><span class="nav-number">4.2.3.1.</span> <span class="nav-text">(1) Exact test (for small data)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-Parallel-exact-test-for-small-data-2"><span class="nav-number">4.2.3.2.</span> <span class="nav-text">(2) Parallel exact test (for small data)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-approximate-test-recommended-for-big-data-2"><span class="nav-number">4.2.3.3.</span> <span class="nav-text">(3) approximate test (recommended for big data)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-Parallel-approximate-test-recommended-for-big-data-2"><span class="nav-number">4.2.3.4.</span> <span class="nav-text">(4) Parallel approximate test (recommended for big data)</span></a></li></ol></li></ol></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2020</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Chao Ning</span>

  
</div>









        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>
  

  
  
    <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.4"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.4"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.4"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.4"></script>



  


  




	





  





  












  





  

  

  

  
  

  

  

  

</body>
</html>
